Feat/image captioner service

ldw129 opened the pull request:
### Summary

<!-- Provide a concise summary of the changes introduced in this PR. Link to relevant issue(s) if applicable. -->

Implement image captioning via Vision Language Model (VLM) functionality for the caption generation feature in #61.

---

### Changes Made

<!-- List out major changes in this PR. Be brief but specific. -->

- Configure docker compose and env files for image captioner service to connect to VLM
- Access openai_client in shared_utils module
- Configure Qwen2.5-VL-7B-Instruct as the primary VLM for image caption generation deployment, optimization and testing

---

### Context / Rationale

<!-- Why are these changes being made? Is it a bug fix, feature, refactor, etc.? Provide context to help reviewers understand your thought process. -->

- Configure docker compose and env files for image captioner service to connect to VLM: To enable image captioner service to call the VLM for the image caption generation feature
- Access openai_client in shared_utils module: To let image captioner service access via API call the shared AsyncOpenAI client instance to load the VLM, for consistency across the services
- Configure Qwen2.5-VL-7B-Instruct as the primary VLM for image caption generation deployment, optimization and testing: To allow the image caption generation feature to generate accurate image captions using VLM, based on the given image received from PDF processor service

---

### Related Docs or References

<!-- Include any design docs, spec references, screenshots, or relevant PRs -->

N.A.

---

### FastAPI Application Checklist (**Delete if PR is not relevant**)

- [x] API follows RESTful principles (nouns in routes, proper use of verbs)
- [x] All endpoints are async and use non-blocking I/O
- [x] `/health` endpoint is implemented and returns 200 OK
- [x] Job results are persisted or recoverable if needed
- [x] API schema (OpenAPI) is exposed and browsable at `/docs` or `/redoc`
- [x] Branch name follows conventions (e.g., `feature/*`, `bugfix/*`) — do **not** use `dev` directly

---

### General Checklist

- [x] I have tested these changes locally
- [x] I have updated relevant documentation or added comments where needed
- [x] I have linked relevant issues and tagged reviewers
- [x] I have followed coding conventions and naming standards

M00dler committed:
feat: added bypass code to add images directly for testing purposes

M00dler committed:
feat: testing of VL model

M00dler committed:
feat: experimented with setting up local VL model

ldw129 committed:
fix: Add pillow and openai libraries to requirement.txt

ldw129 committed:
fix: Update routers for image captioner service

ldw129 committed:
perf: Add new env variables for VLM configuration for image captioner service

- Add new env variables for VLM configuration for image captioner service
- Update OPENAI_BASE_URL in example.env

ldw129 committed:
style: Add newline at end of script as per PEP8 conventions

ldw129 committed:
feat: Implement caption generation using VLM for image captioner service

ldw129 committed:
Merge branch 'dev' into feat/image_captioner

ldw129 committed:
chore: Remove testing scripts and files previously used

- Remove bypass scripts
- Delete model_test.ipynb

ldw129 committed:
fix: Delete model_test.ipynb

ldw129 committed:
fix: Remove image_captioner_service from nginx in docker-compose.yml

ldw129 committed:
fix: Remove image_captioner_service from nginx in docker-compose.yml

ldw129 committed:
chore: Remove bypass scripts

ldw129 committed:
chore: Update caption endpoint to /

ldw129 committed:
chore: Update caption endpoint to /

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,119 @@
+VLM_MODEL = vlm_config.model_name
+
+
+@router.post("/caption", response_model=ImageCaptioningResponse, status_code=200)
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/185a09078af823f5019b1c007dadbd31a02af841)

ldw129 committed:
chore: Remove trailing whitespace in ImageCaptioningResponse class in models/caption

ldw129 committed:
chore: Remove trailing whitespace in ImageCaptioningResponse class in models/caption

ldw129 commented on image_captioner_service/models/caption.py:
```diff
@@ -0,0 +1,13 @@
+class ImageCaptioningResponse(BaseModel):
+    doc_id: str
+    image_id: str 
+    caption: str 
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/f6a1f88291e59ecbc2f3acec69e9aa3a1ca4f3d8)

ldw129 committed:
fix: Update root_path in main

ldw129 commented on image_captioner_service/main.py:
```diff
@@ -0,0 +1,15 @@
+    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s"
+)
+
+app = FastAPI(root_path="/pdf_processor")
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/f8d9a9d18114b9b11f60afdc23c7932adbc81f42)

ldw129 committed:
fix: Use HTTPX AsyncCLient to get image_url in routers/caption

Revert "fix: Use HTTPX AsyncCLient to get image_url in routers/caption"

This reverts commit af19871462bc055fc7753d567629ce8f7054e695.

fix: Use HTTPX AsyncClient to get image_url in routers/caption

ldw129 committed:
fix: Use HTTPX AsyncCLient to get image_url in routers/caption

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,119 @@
+        logger.info(f"Image URL: {request.image_url}")
+        logger.info(f"prompt: {request.prompt}")
+
+        response = await client.get(request.image_url, cast_to=httpx.Response)
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/af19871462bc055fc7753d567629ce8f7054e695)

ldw129 committed:
fix: Create a list containing request.prompt and encoded_image for content to be sent to VLM

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,119 @@
+                "role": "user",
+                "content": encoded_image + "\n\n" + request.prompt # user prompt provided by user or set as fixed by system
+            }
+        ]
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/41211423524ce2c4c1e51da7049e9fb8c5506ba4)

ldw129 commented on image_captioner_service/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+EXPOSE 8000
+
+CMD ["sh", "-c", "PYTHONPATH=/app/image_captioner_service uvicorn image_captioner_service.main:app --host 0.0.0.0 --port 8000"]
```
@gemini-code-assist the `shared_utils` path is already specified in the second COPY instruction.

ldw129 committed:
refactor: Specify exceptions related to the HTTPX AsyncClient in routers/caption

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,123 @@
+        async with httpx.AsyncClient() as http_client:
+            response = await http_client.get(request.image_url, follow_redirects=True)
+
+    except Exception as e:
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/8c69392850d9ef062964e369d4d58b930775094d)

ldw129 committed:
refactor: Specify exception for image fetching in routers/caption

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,123 @@
+
+        logger.info("Successfully downloaded and processed image.")
+
+    except Exception as e:
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/dd7083b167b0ba54bfde2847e517abc451863c42)

ldw129 committed:
refactor: Split into 2 functions for retrieving image and generating image caption in routers/caption

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,123 @@
+    logger.info("Successfully generated caption and sending response.")
+
+    return response_data
+
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/65d8dc04932520c996e29d112ce1e990f8fe2cd1)

ldw129 committed:
refactor: Remove instantiation of static vlm_config classes and call them directly in routers/caption

ldw129 committed:
fix: Update CMD instruction in dockerfile

ldw129 commented on image_captioner_service/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+EXPOSE 8000
+
+CMD ["sh", "-c", "PYTHONPATH=/app/image_captioner_service uvicorn image_captioner_service.main:app --host 0.0.0.0 --port 8000"]
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/8cd697d6efc35ec89f07c801d73367c11dc9cf19)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,136 @@
+# Initialize configuration for VLM
+vlm_config = VLMConfig()
+prompt_templates = PromptTemplates()
+optimizer = CaptionOptimizer()
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/7a36697a88308798de4905fa659bcccdcce119f3)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,136 @@
+        encoded_image = base64.b64encode(image_bytes).decode("utf-8")
+
+        # Prepare messages containing system prompt and encoded image for VLM
+        system_prompt = prompt_templates.get_system_prompt()
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/7a36697a88308798de4905fa659bcccdcce119f3)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,136 @@
+    
+    # Post-process the response
+    if vlm_config.enable_response_post_processing:
+        processed_caption = optimizer.post_process_llm_response(first_choice.message.content)
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/7a36697a88308798de4905fa659bcccdcce119f3)

ldw129 committed:
fix: Strip clean_response when checking if its non-empty in vlm_config

ldw129 commented on image_captioner_service/models/vlm_config.py:
```diff
@@ -0,0 +1,74 @@
+        cleaned_response = '\n'.join(filtered_lines)
+        
+        # Ensure response ends properly
+        if cleaned_response and not cleaned_response.rstrip().endswith(('.', '!', '?', ':')):
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/53434a264eefbb8d8340a4b552b2154d2e81aa2d)

ldw129 committed:
fix: Remove not keyword for current_line_stripped in if block when checking for empty lines and whitespaces

ldw129 committed:
fix: Instantiate VLMConfig() and call PromptTemplates and CaptionOptimizer directly

ldw129 commented on image_captioner_service/models/vlm_config.py:
```diff
@@ -0,0 +1,74 @@
+        for i in range(1, len(lines)):
+            current_line_stripped = lines[i].strip()
+            prev_line_stripped = lines[i-1].strip()
+            if not current_line_stripped or current_line_stripped != prev_line_stripped:
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/c7b633ca95fd5b453774096dbb8dfc6d9b930a3f)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,131 @@
+router = APIRouter(prefix="/caption", tags=["caption"])
+logger = logging.getLogger(__name__)
+
+VLM_MODEL = VLMConfig().model_name
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/514313f946928f43362f51e3516dff9ead36ee84)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,131 @@
+        encoded_image = base64.b64encode(image_bytes).decode("utf-8")
+
+        # Prepare messages containing system prompt and encoded image for VLM
+        system_prompt = PromptTemplates().get_system_prompt()
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/514313f946928f43362f51e3516dff9ead36ee84)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,131 @@
+    
+    # Post-process the response
+    if VLMConfig().enable_response_post_processing:
+        processed_caption = CaptionOptimizer().post_process_llm_response(first_choice.message.content)
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/514313f946928f43362f51e3516dff9ead36ee84)

ldw129 commented:
/gemini review

ldw129 committed:
fix: Remove redundant check for current_line_stripped

ldw129 committed:
chore: Split long code lines into smaller code lines for PEP8

ldw129 commented on image_captioner_service/models/vlm_config.py:
```diff
@@ -0,0 +1,74 @@
+            current_line_stripped = lines[i].strip()
+            prev_line_stripped = lines[i-1].strip()
+            if current_line_stripped or current_line_stripped != prev_line_stripped:
+                filtered_lines.append(lines[i])
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/a70e1652af1dbe38b6728eac1ee74c010541f619)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,132 @@
+    
+    logger.info(f"Received caption from LLM: '{processed_caption}'")
+
+    response_data = ImageCaptioningResponse(doc_id=request.doc_id, image_id=request.image_id, caption=processed_caption)
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/03ef12189537baafeb4160a1200e639601b1b500)

MingJ7 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,141 @@
+        image_bytes = image_bytes.getvalue()
+
+        logger.info("Successfully downloaded and processed image.")
+        return image_bytes
```
Nit: If you are converting the image to base64 later, can the saving step be skipped?

ldw129 replied:
No, doing so will trigger the following error:

`{'error': 'Input Buffer is empty'}`

My understanding is if this line is absent, `image_bytes` would just be empty without the original image, so there would be no image to convert to base64 later.

MingJ7 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,141 @@
+            logger.info("Converting image to RGB mode.")
+            image = image.convert("RGB")
+        
+        # Convert the image to bytes
+        image_bytes = io.BytesIO()
+        image.save(image_bytes, format="JPEG")
+        image_bytes = image_bytes.getvalue()
```
Also, if the file is both RGB and JPEG, why not just return the Binary directly?

ldw129 replied:
Agreed, we should accept images of all modes, including potentially black-white images and not just "RGB".

ldw129 replied:
Ill comment it out as a provision if there is a need to convert all images to a specific image mode in the future.

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/67b56518e96103ca79f7723137624e1f5836aa5e)

MingJ7 commented on nginx/example.env:
```diff
@@ -3,4 +3,5 @@ PDF_EXTRACTOR_URL=http://pdf_extraction_service:8000
 CHAT_URL=http://chat_service:8000
 DOCLING_TRANSLATION_URL=http://docling_translation_service:8000
 EMBEDDER_URL=http://embedder_service:8000
+IMAGE_CAPTIONER_URL=http://image_captioner_service:8000
```
blocking: Image captioner is not a frontend facing service. Please do not forward from nginx.

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/cbb6cc9994966caddf877c48b8fe2bf02ed7df79)

MingJ7 commented on nginx/nginx.conf.template:
```diff
@@ -42,4 +42,8 @@ server {
+
+    location /image_captioner/ {
+        proxy_pass ${IMAGE_CAPTIONER_URL}/;
+    }
```
blocking: Image captioner is not a frontend facing service. Please do not forward from nginx.

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/cbb6cc9994966caddf877c48b8fe2bf02ed7df79)

MingJ7 commented on image_captioner_service/models/caption.py:
```diff
@@ -0,0 +1,13 @@
+from pydantic import BaseModel
+
+class ImageCaptioningRequest(BaseModel):
+    doc_id: str
+    image_id: str
```
Nit: doc_id and image_id are used only for the purpose of logging and returning the values to the user. Consider removing.

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/977d643233c91a8753b6ed79b2dfaf8530864daf)

ldw129 committed:
refactor: Comment out image mode conversion code as a provision for future use if necessary

ldw129 committed:
fix: Remove ImageCaptioningResponse and return caption only

ldw129 committed:
fix: Remove image captioner service from Nginx

Revert "fix: Remove image captioner service from Nginx"

This reverts commit 6c70fdaa994746626953ef8b6c85bbf684bc50ad.

fix: Remove image captioner service from Nginx

ldw129 committed:
refactor: Define http_client at module level in routers/caption

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,136 @@
+        logger.info(f"prompt: {request.prompt}")
+
+        async with httpx.AsyncClient() as http_client:
+            response = await http_client.get(request.image_url, follow_redirects=True)
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/463828c186f2f60aa80e69a81139ea5dd34b90d2)

ldw129 committed:
Revert "fix: Remove ImageCaptioningResponse and return caption only"

This reverts commit 977d643233c91a8753b6ed79b2dfaf8530864daf.

ldw129 committed:
fix: Remove doc_id and image_id from ImageCaptioningResponse model

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,136 @@
+    logger.info(f"Received caption from LLM: '{processed_caption}'")
+    logger.info("Successfully generated caption and sending response.")
+    response_data = {"caption": processed_caption}
+    return response_data
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/08d5624561112b02efee31606ecb2f3ff42db718)

ldw129 committed:
fix: Use environ method instead of getenv to call OPENAI_VLM in vlm_config

ldw129 commented on image_captioner_service/models/vlm_config.py:
```diff
@@ -0,0 +1,74 @@
+    """Configuration class for the optimization of the model's caption generation"""
+    
+    def __init__(self):
+        self.model_name = os.getenv("OPENAI_VLM")
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/54d1b8fe2ed339557d2d5f3a8e4e6b2ea098322c)

ldw129 committed:
Merge branch 'feat/image_captioner' of https://github.com/NotYuSheng/OmniPDF into feat/image_captioner

ldw129 committed:
refactor: Rename function to retrieve image and enhance function logic

ldw129 committed:
chore: Remove unused import in routers/caption

ldw129 commented:
/gemini review

ldw129 committed:
feat: Validate that retrieved image can be opened

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,131 @@
+
+    except (Image.UnidentifiedImageError, IOError) as e:
+        logger.error(f"Error processing image: {e}")
+        raise HTTPException(status_code=500, detail="Failed to process image")
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/0b622bba4652079207866b92d869464c1d58d28a)

ldw129 committed:
fix: Remove unused env variables

ldw129 committed:
Update image_url value to accept image of any format in routers/caption

ldw129 commented on image_captioner_service/example.env:
```diff
@@ -0,0 +1,21 @@
+MINIO_BUCKET=omnifiles
+MINIO_ACCESS_KEY=minioadmin
+MINIO_SECRET_KEY=minioadmin
+REDIS_URL="redis://redis:6379/0"
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/042b8944532f79bac14b96950f5309e9db763c03)

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,131 @@
+                    {"type": "text", "text": request.prompt},
+                    {
+                        "type": "image_url", 
+                        "image_url": {"url": f"data:image/jpeg;base64,{encoded_image}"}
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/299ad6b3d1cabb70bff430abf6850c384f0f501c)

ldw129 committed:
refactor: Refactor logic in post_process_llm_response function

ldw129 commented on image_captioner_service/models/vlm_config.py:
```diff
@@ -0,0 +1,74 @@
+            current_line_stripped = lines[i].strip()
+            prev_line_stripped = lines[i-1].strip()
+            if current_line_stripped != prev_line_stripped:
+                filtered_lines.append(lines[i])
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/beb1c7e453d3b48f65a32651c5d5e25b749bf1e6)

ldw129 committed:
chore: Remove logging statements in routers/caption

ldw129 commented on image_captioner_service/routers/caption.py:
```diff
@@ -0,0 +1,131 @@
+        logger.info(f"doc id: {request.doc_id}")
+        logger.info(f"Image ID: {request.image_id}")
+        logger.info(f"Image URL: {request.image_url}")
+        logger.info(f"prompt: {request.prompt}")
```
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/d09dfb2dc0489fb83783ddb50d42f1627f2987ee)

ldw129 committed:
refactor: Refactor logic for get_image function in routers/caption

NotYuSheng commented on image_captioner_service/example.env:
```diff
@@ -0,0 +1,14 @@
+# Vision Language Model Configuration
+OPENAI_BASE_URL=http://webworkdgx/vllm_qwen2.5/v1 # DGX URL
+OPENAI_VLM=Qwen2.5-VL-7B-Instruct
```
Missing TOKEN

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/2465b978dff0abd3fd9aa78547b0c2a4c66101f1)

NotYuSheng commented on image_captioner_service/example.env:
```diff
@@ -0,0 +1,14 @@
+# Vision Language Model Configuration
+OPENAI_BASE_URL=http://webworkdgx/vllm_qwen2.5/v1 # DGX URL
+OPENAI_VLM=Qwen2.5-VL-7B-Instruct
```
URL and model mismatch, are you sure you've tested this?

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/155/commits/22ddb898f2f322ff14d06d4fbe91ddbb55ce450a)

NotYuSheng reviewed (changes requested):
Fix me please

ldw129 committed:
fix: Add OPENAI_API_KEY and update OPENAI_BASE_URL in env file of image captioner service

ldw129 committed:
refactor: Provide a default user prompt for ImageCaptioningRequest class

ldw129 committed:
fix: Update OPENAI_BASE_URL in env file

ldw129 committed:
fix: Change os.getenv() to os.environ[] for calling the env variables in openai_client

ldw129 committed:
chore: Remove sample image file

ldw129 committed:
refactor: Update VLM's system prompt to include image context

ldw129 committed:
chore: Rearrange module import statements in routers/caption

ldw129 committed:
refactor: Include example caption for each content-specific instruction in VLM's system prompt

NotYuSheng reviewed (approved):
Approved & Merged

Nice job!
